File: typescript/src/game-finder.ts
import * as vdf from '@node-steam/vdf';
import fs from 'fs';
import path from 'path';
import { enumerateValues, HKEY, RegistryValueType } from 'registry-js';
import { Channel, GameFinderOptions } from './game-finder-options';

export class GameFinder {
  public static findGame(options: GameFinderOptions): string {
    try {
      if (!options.onlyEpic) {
        try {
          return GameFinder.getSteamGamePath(options.channel);
        } catch (err) {
          if (options.onlySteam) throw err;
        }
      }
      if (!options.onlySteam)
        return GameFinder.getEpicGamesPath(options.channel);
    }
    catch (err) {
      throw new Error(`Could not find ${options.channel} installation: ${err.message ?? 'unknown error'}. Options=${JSON.stringify(options)}`);
    }
    throw new Error('An unknown error ocurred.');
  }

  private static getRegistryString = (path: string, valueName: string): string | undefined => {
    const ucName = valueName.toLocaleUpperCase();
    for (const value of enumerateValues(HKEY.HKEY_LOCAL_MACHINE, path)) {
      if (value.name.toLocaleUpperCase() === ucName && value.type === RegistryValueType.REG_SZ)
        return value.data;
    }
    return undefined;
  };

  private static findWindowsSteamLibrariesVdf(): string {
    const steamPath = GameFinder.getRegistryString('SOFTWARE\\WOW6432Node\\Valve\\Steam', 'InstallPath');
    if (!steamPath) throw new Error('Steam is not installed properly.');

    // Steam can have multiple library locations, which are in the VDF format in a known location.
    const librariesVdf = path.resolve(steamPath, 'steamapps', 'libraryfolders.vdf');

    return librariesVdf;
  }

  private static findWindowsEpicManifests(): string {
    const appDataPath = GameFinder.getRegistryString('SOFTWARE\\WOW6432Node\\Epic Games\\EpicGamesLauncher', 'AppDataPath');
    if (!appDataPath) {
      throw new Error('Epic Games registry entry not found');
    }

    return path.resolve(appDataPath, 'Manifests');
  }

  private static getEpicGamesPath(channel: Channel): string {
    // This is the Epic Games catalog item id for the Experimental fork of Satisfactory.
    // We could use DisplayName, but that would be less accurate.
    const catalogIds: Record<Channel, string> = {
      experimental: 'ef4a63daa7d4420e91420a72050be89d',
      earlyaccess: 'b915dfe8dcf74770841c82a4162dc954'
    };

    const catalogId = catalogIds[channel];

    // in the Manifests folder, there will be one .item file per installed game.
    const manifestsFolder = this.findWindowsEpicManifests();

    if (!fs.existsSync(manifestsFolder)) {
      throw new Error(`Epic Games manifests directory '${manifestsFolder}' does not exist`);
    }

    for (const itemFileName of fs.readdirSync(manifestsFolder)) {
      if (!itemFileName.endsWith('.item')) continue;
      const itemPath = path.resolve(manifestsFolder, itemFileName);

      // Epic's manifests are in JSON format. Games are keyed by the catalog item id.
      const manifest = JSON.parse(fs.readFileSync(itemPath, { encoding: 'utf8' }));
      if (manifest?.CatalogItemId === catalogId)
        return manifest.InstallLocation;
    }
    throw new Error('Could not find installed game in any Epic Games library');
  }

  private static getSteamGamePath(channel: Channel): string {
    // look in the Windows registry to find Steam's installation path
    const steamAppId = '526870';

    const librariesVdf = this.findWindowsSteamLibrariesVdf();

    if (!fs.existsSync(librariesVdf)) throw new Error(`${librariesVdf} does not exist.`);

    const manifest = vdf.parse(fs.readFileSync(librariesVdf, { encoding: 'utf8' }));
    if (!manifest?.libraryfolders) throw new Error(`No libraryfolders in Steam VDF at ${librariesVdf}`);

    // find the library that contains the Satisfactory game
    for (const k in manifest.libraryfolders) {
      if (!Object.prototype.hasOwnProperty.call(manifest.libraryfolders, k) || typeof manifest.libraryfolders[k] !== 'object')
        continue;

      const library = manifest.libraryfolders[k];

      // sanity checking...
      if (!library || typeof library !== 'object' ||
        !library.path || !fs.existsSync(library.path) ||
        !library.apps || typeof library.apps !== 'object' ||
        // if the library contains the game, this will be present.
        library.apps[steamAppId] === undefined) {
        continue;
      }

      const steamappsDir = path.resolve(library.path, 'steamapps');

      // Steam thinks the game is installed, so look for the game's local install path in its VDF manifest.
      const appManifestPath = path.resolve(steamappsDir, `appmanifest_${steamAppId}.acf`);
      if (!fs.existsSync(appManifestPath)) {
        continue;
      }

      // the manifest exists, so parse it
      const appManifest = vdf.parse(fs.readFileSync(appManifestPath, { encoding: 'utf8' })) as SteamAppManifest;

      // verify that this is the correct channel
      if (getSteamChannel(appManifest) !== channel) {
        continue;
      }

      const installDir = appManifest?.AppState?.installdir;
      if (!installDir) {
        continue;
      }

      // Steam thinks the game is installed in this library, but sanity check that it actually still exists.
      const installPath = path.resolve(steamappsDir, 'common', installDir);
      if (!fs.existsSync(installPath)) continue;

      return installPath;
    }

    throw new Error('Could not find installed game in any Steam library');
  }
}

interface SteamAppManifest {
  AppState: {
    appid: string;
    installdir?: string;
    UserConfig?: {
      betakey?: string
    }
  };
}

const getSteamChannel = (appManifest: SteamAppManifest): Channel => {
  return appManifest.AppState.UserConfig?.betakey === 'experimental'
    ? 'experimental'
    : 'earlyaccess';
};

File: dotnet/src/SatisfactoryExtractor/GameFinder.cs
namespace Rosser.SatisfactoryExtractor;

using Gameloop.Vdf;
using Gameloop.Vdf.JsonConverter;
using Microsoft.Win32;
using Newtonsoft.Json.Linq;
using Rosser.SatisfactoryExtractor.Models;
using System.Text.Json;

#pragma warning disable IDE0079 // Remove unnecessary suppression
[System.Diagnostics.CodeAnalysis.SuppressMessage("Interoperability", "CA1416:Validate platform compatibility", Justification = "Has runtime checks")]
#pragma warning restore IDE0079 // Remove unnecessary suppression
public class GameFinder
{
    public static string FindGame(GameFinderOptions options)
    {
        if (OsHelper.IsLinux)
        {
            throw new NotSupportedException("Not supported in Linux.");
        }

        try
        {
            if (!options.OnlyEpic)
            {
                try
                {
                    return GetSteamGamePath(options.Channel);
                }
                catch (Exception)
                {
                    if (options.OnlySteam)
                    {
                        throw;
                    }
                }
            }

            if (!options.OnlySteam)
            {
                return GetEpicGamesPath(options.Channel);
            }
        }
        catch (GameFinderException)
        {
            throw;
        }
        catch (Exception ex)
        {
            throw new GameFinderException($"Could not find {options.Channel} installation: {ex.Message ?? "unknown error"}. Options={options}", ex);
        }

        throw new Exception("An unknown error ocurred.");
    }

    private static string? GetRegistryString(string path, string valueName)
    {
        using RegistryKey? key = Registry.LocalMachine.OpenSubKey(path);
        return key?.GetValueKind(valueName) == RegistryValueKind.String
            ? key.GetValue(valueName) as string
            : null;
    }

    private static string FindWindowsSteamLibrariesVdf()
    {
        string? steamPath = GetRegistryString(@"SOFTWARE\WOW6432Node\Valve\Steam", "InstallPath");
        if (steamPath is null)
        {
            throw new GameFinderException("Steam is not installed properly.");
        }

        // Steam can have multiple library locations, which are in the VDF format in a known location.
        string librariesVdf = Path.Combine(steamPath, "steamapps", "libraryfolders.vdf");

        return librariesVdf;
    }

    private static string FindWindowsEpicManifests()
    {
        string? appDataPath = GetRegistryString(@"SOFTWARE\WOW6432Node\Epic Games\EpicGamesLauncher", "AppDataPath");
        return appDataPath is null
            ? throw new GameFinderException("Epic Games registry entry not found")
            : Path.Combine(appDataPath, "Manifests");
    }

    private static string GetEpicGamesPath(Channel channel)
    {
        // This is the Epic Games catalog item id for the Experimental fork of Satisfactory.
        // We could use DisplayName, but that would be less accurate.
        string catalogId = channel switch
        {
            Channel.Experimental => "ef4a63daa7d4420e91420a72050be89d",
            Channel.EarlyAccess => "b915dfe8dcf74770841c82a4162dc954",
            _ => throw new NotSupportedException("Invalid channel")
        };

        // in the Manifests folder, there will be one .item file per installed game.
        string manifestsFolder = FindWindowsEpicManifests();

        if (!Directory.Exists(manifestsFolder))
        {
            throw new DirectoryNotFoundException($"Epic Games manifests directory \"{manifestsFolder}\" does not exist");
        }

        foreach (string itemFileName in Directory.GetFiles(manifestsFolder))
        {
            if (!itemFileName.EndsWith(".item"))
            {
                continue;
            }

            string itemPath = Path.Combine(manifestsFolder, itemFileName);

            // Epic's manifests are in JSON format. Games are keyed by the catalog item id.
            using StreamReader reader = new(itemPath);
            EpicManifest? manifest = JsonSerializer.Deserialize<EpicManifest>(reader.ReadToEnd());
            if (manifest?.CatalogItemId == catalogId)
            {
                return manifest.InstallLocation;
            }
        }

        throw new GameFinderException("Could not find installed game in any Epic Games library");
    }

    private static string GetSteamGamePath(Channel channel)
    {
        // look in the Windows registry to find Steam's installation path
        const string steamAppId = "526870";

        string librariesVdf = FindWindowsSteamLibrariesVdf();

        if (!File.Exists(librariesVdf))
        {
            throw new FileNotFoundException(librariesVdf);
        }

        JObject manifest = VdfConvert.Deserialize(File.ReadAllText(librariesVdf)).Value.ToJson() as JObject
            ?? throw new GameFinderException($"Could not deserialize Steam VDF at {librariesVdf}");

        // find the library that contains the Satisfactory game
        foreach (JProperty prop in manifest.Children().Cast<JProperty>())
        {
            var library = prop.Value as JObject;
            string? libraryPath = library?["path"]?.Value<string>();
            // sanity checking...
            if (library is null ||
                libraryPath is null || !Directory.Exists(libraryPath) ||
                library.GetValue("apps") is not JObject apps ||
                // if the library contains the game, this will be present.
                !apps.ContainsKey(steamAppId))
            {
                continue;
            }

            string steamappsDir = Path.Combine(libraryPath, "steamapps");

            // Steam thinks the game is installed, so look for the game's local install path in its VDF manifest.
            string appManifestPath = Path.Combine(steamappsDir, $"appmanifest_{steamAppId}.acf");
            if (!File.Exists(appManifestPath))
            {
                continue;
            }

            // the manifest exists, so parse it
            if (VdfConvert.Deserialize(File.ReadAllText(appManifestPath)).Value.ToJson() is not JObject appManifest)
            {
                continue;
            }

            // verify that this is the correct channel
            Channel appChannel = appManifest["UserConfig"]?["betakey"]?.Value<string>() switch
            {
                "experimental" => Channel.Experimental,
                _ => Channel.EarlyAccess
            };
            if (appChannel != channel)
            {
                continue;
            }

            string? installDir = appManifest["installdir"]?.Value<string>();
            if (installDir is null)
            {
                continue;
            }

            // Steam thinks the game is installed in this library, but sanity check that it actually still exists.
            string installPath = Path.Combine(steamappsDir, "common", installDir);
            if (!Directory.Exists(installPath))
            {
                continue;
            }

            return installPath;
        }

        throw new GameFinderException("Could not find installed game in any Steam library");
    }
}